book/trees/avl_tree.py

```diff
@@ -133,29 +125,50 @@ def update_balance(self,node):
 to point to the new root. Finally we set the parent of
 the old root to be the new root. This is a lot of complicated
 bookkeeping, so we encourage you to trace through this function while
-looking at the diagram above. The `rotate_right` method
-is symmetrical to `rotate_left` so we will leave it to you to implement the
-code for `rotate_right`.
-
-```python
-def rotate_left(self, rotation_root):
-    new_root = rotation_root.right
-    rotation_root.right = new_root.left
-    if new_root.left != None:
-        new_root.left.parent = rotation_root
-    new_root.parent = rotation_root.parent
-    if rotation_root.is_root():
-        self.root = new_root
-    else:
-        if rotation_root.is_left_child():
-            rotation_root.parent.left = new_root
+looking at the diagram above.
+
+"""
+    def rotate_left(self, rotation_root):
+        new_root = rotation_root.right
+        rotation_root.right = new_root.left
+        if new_root.left is not None:
+            new_root.left.parent = rotation_root
+        new_root.parent = rotation_root.parent
+        if not rotation_root.parent:
+            self.root = new_root
         else:
-            rotation_root.parent.right = new_root
-    new_root.left = rotation_root
-    rotation_root.parent = new_root
-    rotation_root.balance_factor = rotation_root.balance_factor + 1 - min(new_root.balance_factor, 0)
-    new_root.balance_factor = new_root.balance_factor + 1 + max(rotation_root.balance_factor, 0)
-```
+            if rotation_root.is_left_child():
+                rotation_root.parent.left = new_root
+            else:
+                rotation_root.parent.right = new_root
+        new_root.left = rotation_root
+        rotation_root.parent = new_root
+        rotation_root.balance_factor = \
+            rotation_root.balance_factor + 1 - min(new_root.balance_factor, 0)
+        new_root.balance_factor = \
+            new_root.balance_factor + 1 + max(rotation_root.balance_factor, 0)
+
+    def rotate_right(self, rotation_root):
```

Comment:
I couldn't find a great way to factor out the common logic between `rotate_left` and `rotate_right` without making the code harder to read :(